### PR TITLE
fix(domRenderer) ng-leave doesn’t works

### PR DESCRIPTION
**Pull request by breath103**
i was testing NgIf / ng-animate, and ng-enter was working perfectly but ng-leave class wasn't even show up 

while checking DomRenderer codes, i found a bug which is `detachView` method removing DOM before `animateNodeLeave`.

and if you see the `attachView` in other hands, https://github.com/angular/angular/blob/0ae77753f34862a6f584e131af1630dbd588b5de/modules/angular2/src/platform/dom/dom_renderer.ts#L139
this basically delegating all the DOM add to animateNodeEnter.

and eventually `animateNodeLeave` calls `DOM.remove`, regardless using ng-animate or not.

we're already using angular2 BETA for almost production level product, so that this issue is pretty much urgent since we can't add any animation for leaving at this point haha
(i know there is a long plan for ng2-animate, but still)

so let me know how you guys think. thanks

### CODE REVIEW

**Comment by breath103** [comment]
I signed it!

**Comment by tbosch** [comment]
Hi,
thanks, this is is legit. Could you create a test for this?

**Comment by breath103** [comment]
@tbosch can you tell me where i can start with this testing?
i did consider about testing this, but skipped it in initial commit mainly because i couldn't find any existing unit test or spec that i can start with.
from ```/Users/suzy/Works/angular/modules/angular2/test/platform/dom/dom_renderer_integration_spec.ts``` 
maybe? 

(After renaming it to just dom_renderer_spec.ts)

**Comment by tbosch** [comment]
Have a look at one of the tests here: https://github.com/angular/angular/blob/master/modules/angular2/test/core/linker/integration_spec.ts

File: modules/angular2/test/web_workers/worker/renderer_integration_spec.ts [hunk 1]
```diff
@@ -193,29 +193,105 @@ export function main() {
...
+                fixture.debugElement.componentInstance.ctxBoolProp = false;
+                fixture.detectChanges();
+                expect(rootEl).toHaveText('');
+
+                async.done();
+              });
+        })
+      );
+
+      fdescribe("with ng-animate", () => {
```
Comment (gkalpak) [inline modules/angular2/test/web_workers/worker/renderer_integration_spec.ts]: _fdescribe_

**Comment by breath103** [comment]
oh it's still in progress. initially i thought this would be pretty straight forward but it's not since 
a) test / build is incredibly slow in my Macbook Air,
b) async testing is kinda not working with css animation completion..

this might gonna take a while. i'll let you know

**Comment by breath103** [comment]
@tbosch

**Comment by breath103** [comment]
@gkalpak ok so now it's only failing on Dart Lint, which i have no idea. can you check?

**Comment by breath103** [comment]
@gkalpak any update?

**Comment by Kriskit** [comment]
Be good to get this sorted and merged in :+1:

**Comment by tbosch** [comment]
We will land a more flexible version of animations soon. Closing this as the code will be replaced then.
/cc @matsko

**Comment by breath103** [comment]
@tbosch is there any milestone or clear plan for animation fix in real soon? otherwise, we're kinda really urgently need this haha like i mentioned we're using angular2 on production and we can't add any animation with ng-animate because of this

i totally understand this gonna be legacy code soon, but until than can we just merge this first?

**Comment by BojDom** [comment]
So where is the progress of this?

**Comment by rhysburnie** [comment]
It would be great if you can have the option to leave the old route outlet visible so you can animate the new over the top of it and then remove the old after the new transition has completed. 
This kind of behaviour is often requested building apps etc.

The kinds of transitions a designer would expect possible are:

* straight cut (all ready possible)
* transition out (1st) / transition in (2nd)
  * simultaneous
  * one after the another
* transition in (1st) / transition out (2nd)
  * simultaneous
  * one after the another

With those hooks you could do any effect.

Also if you chose to do the transition with something like GSAP being able to start the next transition hook at the competition of the animation would be ideal.

Obviously standard behaviour would be simple classname changes for css transitions

**Comment by matsko** [comment]
Note that now we have the new API in place (which has working support for [leave animations](https://angular.io/docs/ts/latest/guide/animations.html#!#example-entering-and-leaving)). Routes are currently not animateable, but we do have solution that should land next week.

**Comment by JoshuaNovak919** [comment]
Any update on being able to animate out? I'm applying a css class to an element on `ngOnInit`, but I want the previous page to stay until i've animated it out before removing it.

**Pull request closed without merging**